clients/drcachesim/tools/invariant_checker.cpp

```diff
@@ -1029,15 +1062,37 @@ invariant_checker_t::parallel_shard_memref(void *shard_data, const memref_t &mem
                      shard->file_type_)) {
             if (type_is_read(memref.data.type)) {
                 // Skip D-filtered traces which don't have every load or store records.
-                report_if_false(shard, shard->expected_read_records_ > 0,
-                                "Too many read records");
+                report_if_false(
+                    shard,
+                    shard->expected_read_records_ >
+                        0 IF_X86(
+                            // iret pops bunch of data from the stack at interrupt
+                            // returns. Read count cannot be determined using the
+                            // decoder. System call trace templates collected on
```

Comment:
This doesn't sound right: the read count is statically determinable as otherwise DR's handling of iret would fail trying to find the retaddr and other args.  The decoder has a special category for it `TYPE_INDIR_VAR_REG_SIZEx3x5` so the size should be there.

Reply:
For the following iret that is marked as having a 40 byte read, I'm seeing variable count of reads split across 8/4 byte reads, totalling to 80 or 48 bytes.

```
iret   %rsp (%rsp)[40byte] -> %rsp
```

Reply:
> as otherwise DR's handling of iret would fail

Perhaps an iret in user-space is different than one in the kernel? See https://www.felixcloutier.com/x86/iret:iretd:iretq which notes multiple different iret behavior in protected mode.

Note that the relaxation I introduced in the invariant check is only for the kernel part of the trace.

Reply:
So a different-privilege-level return does more work: so the IR needs an update then instead of saying it's always 40 bytes.  File an issue and cite it here?